Better edit (replace handling)

BillCarsonFr opened the pull request:
<!-- Please read [CONTRIBUTING.md](https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md) before submitting your pull request -->
 
## Type of change

- [ ] Feature
- [ ] Bugfix
- [x] Technical
- [ ] Other :

## Content

<!-- Describe shortly what has been changed -->
This PR is improving the handling of Edits, and enforcing edit validation rules.
- Edit validation: see `EventEditValidator` ensures the rules as [per spec](https://spec.matrix.org/v1.4/client-server-api/#validity-of-replacement-events) (recently updated)
- Added proper unit tests
- Persistence layer changes:
     - Session data migration needed (v43)
     - We now store a reference of the edit event instead of the content to allow more flexibility when validated event; `lastEdit` instead of `LastContent` (as sender/types are also needed for validation)
     - Late decryption supported with revalidation of edit chain

- Fixed a problem in Aggregator with `process` calls beeing suspend, which is a problem as it's called within a realm transaction and realm instance is locked to a thread. All `asyncTransaction` helpers have been modified to make the `block`param not suspendable

Handling of server side aggregation is still not handled as there is [ongoing discussions](https://github.com/matrix-org/matrix-spec/issues/1299) regarding that

## Motivation and context

<!-- Provide link to the corresponding issue if applicable or explain the context -->
https://github.com/vector-im/element-android/issues/7430


## Screenshots / GIFs
N/A

<!--
|Before|After|
|-|-|
|||
 -->

## Tests

Try some edits, polls, location sharing 

## Tested devices

- [ ] Physical
- [x] Emulator
- OS version(s):

## Checklist

<!-- Depending on the Pull Request content, it can be acceptable if some of the following checkboxes stay unchecked. -->

- [x] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [x] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/events/model/AggregatedRelations.kt:
```diff
@@ -50,5 +50,6 @@ import com.squareup.moshi.JsonClass
 data class AggregatedRelations(
         @Json(name = "m.annotation") val annotations: AggregatedAnnotation? = null,
         @Json(name = "m.reference") val references: DefaultUnsignedRelationInfo? = null,
+        @Json(name = "m.replace") val replaces: AggregatedReplace? = null,
```
I think we may need to update the related Kdoc. And it seems the preview of this doc in the IDE is not working, the `<code></code> `seems to not be well rendered.

BillCarsonFr replied:
fixed, should use ```

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/events/model/ValidDecryptedEvent.kt:
```diff
@@ -0,0 +1,53 @@
+        val algorithm: String,
+)
+
+fun Event.toValidDecryptedEvent(): ValidDecryptedEvent? {
```
I think we should try to add unit tests on this method.

BillCarsonFr replied:
added

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/timeline/TimelineEvent.kt:
```diff
@@ -142,13 +143,21 @@ fun TimelineEvent.getEditedEventId(): String? {
-        in EventType.BEACON_LOCATION_DATA -> (annotations?.editSummary?.latestContent ?: root.getClearContent()).toModel<MessageBeaconLocationDataContent>()
-        else -> (annotations?.editSummary?.latestContent ?: root.getClearContent()).toModel()
+        // XXX
+        // Polls/Beacon are not message contents like others as there is no msgtype subtype to discriminate moshi parsing
```
Thanks for adding this description comment! Indeed, when adding the implementation for beacon events, I was concerned about that. But I couldn't find any other solution since the timeline mecanism seems to rely a lot on `MessageContent`... Do you know if there is any example of events like these which are correctly implemented?

BillCarsonFr replied:
I think that they should not be at all MessageContent from the beginning. could use MessageRelationContent though?
Where is it causing problem?

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/mapper/EventAnnotationsSummaryMapper.kt:
```diff
@@ -36,13 +37,22 @@ internal object EventAnnotationsSummaryMapper {
                 editSummary = annotationsSummary.editSummary
-                        ?.let {
-                            val latestEdition = it.editions.maxByOrNull { editionOfEvent -> editionOfEvent.timestamp } ?: return@let null
+                        ?.let { summary ->
```
I think we should take the occasion to introduce a dedicated `EditAggregatedSummaryMapper` (as a classical class and not an `Object`). We could then add unit tests on it.

BillCarsonFr replied:
Extracted as object and added tests

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/mapper/EventAnnotationsSummaryMapper.kt:
```diff
@@ -36,13 +37,22 @@ internal object EventAnnotationsSummaryMapper {
+                             * if two or more replacement events have identical origin_server_ts,
+                             * the event with the lexicographically largest event_id is treated as more recent.
+                             */
+                            val latestEdition = summary.editions.sortedWith(compareBy<EditionOfEvent> { it.timestamp }.thenBy { it.eventId })
```
I am not sure the implementation is correct. I think the natural order of sorting will be the increasing order and here we would like decreasing order so that we get the most recent first. And I am also not sure about the default sort for strings. 
This is one of the reasons, I strongly think we need unit tests to validate the implementation.

What do you think?

BillCarsonFr replied:
added tests

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/mapper/EventAnnotationsSummaryMapper.kt:
```diff
@@ -36,13 +37,22 @@ internal object EventAnnotationsSummaryMapper {
+                             */
+                            val latestEdition = summary.editions.sortedWith(compareBy<EditionOfEvent> { it.timestamp }.thenBy { it.eventId })
+                                    .lastOrNull() ?: return@let null
+                            // get the event and validate?
```
Why is there this comment? Is there a validation step missing?

BillCarsonFr replied:
no should be validated before, or for late decryption after

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/migration/MigrateSessionTo008.kt:
```diff
@@ -25,11 +25,11 @@ internal class MigrateSessionTo008(realm: DynamicRealm) : RealmMigrator(realm, 8
     override fun doMigrate(realm: DynamicRealm) {
         val editionOfEventSchema = realm.schema.create("EditionOfEvent")
-                .addField(EditionOfEventFields.CONTENT, String::class.java)
+                .addField("content"/**EditionOfEventFields.CONTENT*/, String::class.java)
```
Maybe we should add a comment about this commented code? I guess you wanted to keep the previous fields in comments for future reference?

BillCarsonFr replied:
Kept it to show that it was from an added then deleted class, ok to remove though

mnaturel commented on vector/src/main/java/im/vector/app/core/extensions/TimelineEvent.kt:
```diff
@@ -40,7 +41,8 @@ fun TimelineEvent.getVectorLastMessageContent(): MessageContent? {
     return when (root.getClearType()) {
         VoiceBroadcastConstants.STATE_ROOM_VOICE_BROADCAST_INFO -> {
-            (annotations?.editSummary?.latestContent ?: root.getClearContent()).toModel<MessageVoiceBroadcastInfoContent>()
+            (annotations?.editSummary?.latestEdit?.getClearContent()?.toModel<MessageContent>().toContent().toModel<MessageVoiceBroadcastInfoContent>()
```
I wonder if we could replace this by:
`getLastEditNewContent().toModel<MessageVoiceBroadcastInfoContent>()`? The difference is that in this case we would use the `newContent` field instead of `toContent()` extension method.

mnaturel commented on vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MessageItemFactory.kt:
```diff
@@ -158,7 +158,7 @@ class MessageItemFactory @Inject constructor(
         event.root.eventId ?: return null
         roomId = event.roomId
-        val informationData = messageInformationDataFactory.create(params)
+        val informationData = messageInformationDataFactory.create(params, params.event.annotations?.editSummary?.latestEdit)
```
I am not sure it is needed to add the `lastEdit` as a new argument. Since it is already contained in the first argument, I think we could retrieve it internally inside the create method of the `MessageInformationDataFactory`.

mnaturel commented on vector/src/main/java/im/vector/app/features/home/room/detail/timeline/helper/MessageInformationDataFactory.kt:
```diff
@@ -55,7 +57,7 @@ class MessageInformationDataFactory @Inject constructor(
 ) {
 
-    fun create(params: TimelineItemFactoryParams): MessageInformationData {
+    fun create(params: TimelineItemFactoryParams, lastEdit: Event? = null): MessageInformationData {
```
I am not sure it is needed to add the `lastEdit` as a new argument. Since it is already contained in the first argument, I think we could retrieve it easily inside this method.

BillCarsonFr replied:
good catch

mnaturel commented on matrix-sdk-android/src/test/java/org/matrix/android/sdk/test/fakes/FakeRealmConfiguration.kt:
```diff
@@ -33,9 +33,9 @@ internal class FakeRealmConfiguration {
-        val transaction = slot<suspend (Realm) -> T>()
-        coEvery { awaitTransaction(instance, capture(transaction)) } coAnswers {
-            secondArg<suspend (Realm) -> T>().invoke(realm)
+        val transaction = slot<(Realm) -> T>()
```
I realized the capture of the second argument is in fact not necessary when mocking the answer, so we could simplify this by:
```
fun <T> givenAwaitTransaction(realm: Realm) {
    coEvery { awaitTransaction(instance, any<(Realm) -> T>()) } answers {
        secondArg<(Realm) -> T>().invoke(realm)
    }
}
```

BillCarsonFr replied:
indeed

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:
```diff
@@ -72,20 +72,23 @@ internal class EventRelationsAggregationProcessor @Inject constructor(
             EventType.REDACTION,
             EventType.REACTION,
+            // The aggregator handles verification events but just to render tiles in the timeline
+            // It's not participating in verfication it's self, just timeline display
```
```suggestion
            // It's not participating in verification itself, just timeline display
```

BillCarsonFr replied:
done

mnaturel commented on matrix-sdk-android/src/test/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactoryTests.kt:
```diff
@@ -226,16 +224,14 @@ class LocalEchoEventFactoryTests {
                 eventId = AN_EVENT_ID,
                 displayIndex = 0,
                 senderInfo = SenderInfo(A_USER_ID_1, A_USER_ID_1, true, null),
-                annotations = if (textContent != null) {
```
The remove of these lines means the `editSummary` was not used inside the related tests, am I right?

BillCarsonFr replied:
I don't know why it was used at all like that

mnaturel commented on vector/src/main/java/im/vector/app/features/home/room/detail/timeline/helper/MessageInformationDataFactory.kt:
```diff
@@ -72,8 +74,14 @@ class MessageInformationDataFactory @Inject constructor(
-        val e2eDecoration = getE2EDecoration(roomSummary, event)
-
+        val e2eDecoration = getE2EDecoration(roomSummary, lastEdit ?: event.root)
+        val senderId = if (event.isEncrypted()) {
```
Maybe we should extract this inside a dedicated private method such as `getSenderId()` to reduce complexity of the `create` method?

BillCarsonFr replied:
done

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventEditValidator.kt:
```diff
@@ -0,0 +1,117 @@
+                return EditValidity.Invalid("replacement event must have an m.new_content property")
+            }
+        } else {
+            if (originalEvent.content.toModel<MessageContent>()?.relatesTo?.type == RelationType.REPLACE) {
```
Should we use `getRelationContent` extension as we may miss some specific cases?
`originalEvent.getRelationContent()?.type == RelationType.REPLACE`

BillCarsonFr replied:
good point

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventEditValidator.kt:
```diff
@@ -0,0 +1,117 @@
+        }
+        // check it's from same sender
+
+        if (originalEvent.isEncrypted()) {
```
Should we encapsulate the 2 different cases (encrypted vs non encrypted) in 2 different sub methods to reduce the size of this method?

BillCarsonFr replied:
I see the warning, but I am not sure it makes it easier to read tbh

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventEditValidator.kt:
```diff
@@ -0,0 +1,117 @@
+            val originalCryptoSenderId = cryptoStore.deviceWithIdentityKey(originalDecrypted.cryptoSenderKey)?.userId
+            val editCryptoSenderId = cryptoStore.deviceWithIdentityKey(replaceDecrypted.cryptoSenderKey)?.userId
+
+            if (originalDecrypted.clearContent.toModel<MessageContent>()?.relatesTo?.type == RelationType.REPLACE) {
```
Should we use `getRelationContent` extension as we may miss some specific cases?
`originalEvent.getRelationContent()?.type == RelationType.REPLACE`

And looking at the implementation, I see the encrypted case is handled differently. In the extension method, we have:
`content.toModel<EncryptedEventContent>()?.relatesTo`
What is the correct implementation?

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventEditValidator.kt:
```diff
@@ -0,0 +1,117 @@
+            }
+
+            // check the sender
+            if (originalEvent.senderId != replaceEvent.senderId) {
```
Should we handle the case where both `senderId` are `null`?

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventEditValidator.kt:
```diff
@@ -0,0 +1,117 @@
+                // let valid and UI should display send by deleted device warning?
+                val bestEffortOriginal = originalCryptoSenderId ?: originalEvent.senderId
+                val bestEffortEdit = editCryptoSenderId ?: replaceEvent.senderId
+                if (bestEffortOriginal != bestEffortEdit) {
```
Should we handle the case where `bestEffortOriginal` and `bestEffortEdit` are both `null`?

BillCarsonFr replied:
senderId is null only for the json parsing, but it should be always set. I think we might need some sort of ValidEvent maybe to avoid all these nullable?

mnaturel commented on matrix-sdk-android/src/test/java/org/matrix/android/sdk/internal/session/room/EditValidationTest.kt:
```diff
@@ -0,0 +1,366 @@
+import org.matrix.android.sdk.api.session.events.model.EventType
+import org.matrix.android.sdk.internal.crypto.store.IMXCryptoStore
+
+class EditValidationTest {
```
I think we should name it `EventEditValidatorTest`. This is the convention for unit tests. And using this naming allows navigation in the IDE from a component by using "Right click -> Go to -> Test".

BillCarsonFr replied:
renamed

mnaturel commented on matrix-sdk-android/src/test/java/org/matrix/android/sdk/internal/session/room/EditValidationTest.kt:
```diff
@@ -0,0 +1,366 @@
+
+    @Test
+    fun `edit should be valid`() {
+        val mockCryptoStore = mockk<IMXCryptoStore>()
```
Usually, we try to have 3 seperated blocks in a test method to easily identify the given, when, then blocks. Like this:
```
val mockCryptoStore = mockk<IMXCryptoStore>()
val validator = EventEditValidator(mockCryptoStore)

val validity = validator.validateEdit(mockTextEvent, mockEdit)

validity shouldBeInstanceOf EventEditValidator.EditValidity.Valid::class
```
Maybe we should follow the same pattern on these tests?
On my side I also add `// Given // When // Then` comments before each line to make it clear.

mnaturel commented on matrix-sdk-android/src/test/java/org/matrix/android/sdk/internal/session/room/EditValidationTest.kt:
```diff
@@ -0,0 +1,366 @@
+
+    @Test
+    fun `original event and replacement event must have the same sender`() {
+        val mockCryptoStore = mockk<IMXCryptoStore>()
```
Like for other fake interfaces, I think we should create a new reusable Fake class for `IMXCryptoStore`.

mnaturel commented on matrix-sdk-android/src/test/java/org/matrix/android/sdk/internal/session/room/EditValidationTest.kt:
```diff
@@ -0,0 +1,366 @@
+
+class EditValidationTest {
+
+    private val mockTextEvent = Event(
```
Maybe we should move those objects as well as the ones for encrypted case in the `test.fixtures` package.

mnaturel commented on matrix-sdk-android/src/test/java/org/matrix/android/sdk/internal/session/room/EditValidationTest.kt:
```diff
@@ -0,0 +1,366 @@
+                        mockEdit.copy(stateKey = "")
+                ) shouldBeInstanceOf EventEditValidator.EditValidity.Invalid::class
+
+        validator
```
I think it should be:

```
validator
                .validateEdit(
                        encryptedEvent.copy(stateKey = ""),
                        encryptedEditEvent
                ) shouldBeInstanceOf EventEditValidator.EditValidity.Invalid::class
```

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:
```diff
@@ -113,21 +140,17 @@ internal class EventRelationsAggregationProcessor @Inject constructor(
-                                    TimelineEventEntity.where(realm, roomId = roomId, eventId = event.eventId ?: "").findAll()
-                                            ?.forEach { tet -> tet.annotations = it }
-                                }
+                        // XXX do something for aggregated edits?
```
Reading the comments, Is there something to be done there?

BillCarsonFr replied:
yes for later. support for all the edge cases for edits is not yet there and we don't yet support server side aggregation (some spec concerns)

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:
```diff
@@ -142,73 +165,30 @@ internal class EventRelationsAggregationProcessor @Inject constructor(
-//                                     TimelineEventEntity.where(realm, roomId = roomId, eventId = event.eventId ?: "").findAll()
-//                                             ?.forEach { tet -> tet.annotations = it }
-//                                 }
+//                    // Relation type is in clear, it might be possible to do some things?
```
Is it intended this code is commented?

BillCarsonFr replied:
added back the logs

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:
```diff
@@ -305,11 +281,10 @@ internal class EventRelationsAggregationProcessor @Inject constructor(
                         editSummary.editions.add(
                                 EditionOfEvent(
-                                        senderId = event.senderId ?: "",
                                         eventId = event.eventId,
```
```suggestion
                                        eventId = eventId,
```

mnaturel replied:
Have you seen this suggestion?

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:
```diff
@@ -334,9 +309,8 @@ internal class EventRelationsAggregationProcessor @Inject constructor(
                 existingSummary.editions.add(
                         EditionOfEvent(
-                                senderId = event.senderId ?: "",
                                 eventId = event.eventId,
```
```suggestion
                                eventId = eventId,
```

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:
```diff
@@ -369,8 +343,7 @@ internal class EventRelationsAggregationProcessor @Inject constructor(
     private fun handleThreadSummaryEdition(
-            editedEvent: EventEntity?,
-            replaceEvent: TimelineEventEntity?,
+            editedEvent: EventEntity?, replaceEvent: TimelineEventEntity?,
```
Maybe it is better to have on argument per line for consistency?

BillCarsonFr replied:
dones

mnaturel reviewed (changes requested):
I have posted some comments concerning some implementation details that should be investigated I guess.

After some tests, here the issues I found:
- in non encrypted room, the last edited content is not rendered in the timeline for a text message for example. This is always the original content which is rendered
- in encrypted room, not really sure but I think we might not more render the local echo of the edited content in the timeline as I have seen some short delay before rendering (it may wait for the remote content before rendering).

I have not tested the updated of the app for migration of the DB. If not already done, I think the migration should be tested before merge.

BillCarsonFr commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/events/model/ValidDecryptedEvent.kt:
```diff
@@ -0,0 +1,53 @@
+        val algorithm: String,
+)
+
+fun Event.toValidDecryptedEvent(): ValidDecryptedEvent? {
```
added

BillCarsonFr commented on vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/TimelineItemFactoryParams.kt:
```diff
@@ -19,10 +19,12 @@ package im.vector.app.features.home.room.detail.timeline.factory
 
 data class TimelineItemFactoryParams(
         val event: TimelineEvent,
+        val lastEdit: Event? = null,
```
it's used now

BillCarsonFr commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:
```diff
@@ -72,20 +72,23 @@ internal class EventRelationsAggregationProcessor @Inject constructor(
             EventType.KEY_VERIFICATION_ACCEPT,
             EventType.KEY_VERIFICATION_START,
             EventType.KEY_VERIFICATION_MAC,
             // TODO Add ?
```
yes

BillCarsonFr commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:
```diff
@@ -274,23 +251,22 @@ internal class EventRelationsAggregationProcessor @Inject constructor(
-        val newContent = content.newContent ?: return
-
-        // Check that the sender is the same
+        val targetEventId = relatedEventId ?: return // ?: content.relatesTo?.eventId ?: return
```
done

mnaturel commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/events/model/EventExt.kt:
```diff
@@ -0,0 +1,46 @@
+ * limitations under the License.
+ */
+
+package org.matrix.android.sdk.api.session.events.model
```
Were you aware there was an existing `EventExt` in `the org.matrix.android.sdk.internal.session.events` package?

BillCarsonFr replied:
Yes but it's internal, and toValidate is API

mnaturel replied:
Okay I see. Thanks for explaining.

mnaturel commented on matrix-sdk-android/src/test/java/org/matrix/android/sdk/internal/session/event/ValidDecryptedEventTest.kt:
```diff
@@ -0,0 +1,134 @@
+
+class ValidDecryptedEventTest {
+
+    val fakeEvent = Event(
```
I think we should declare the variable as `private`.

BillCarsonFr replied:
done

mnaturel commented on matrix-sdk-android/src/test/java/org/matrix/android/sdk/internal/session/event/ValidDecryptedEventTest.kt:
```diff
@@ -0,0 +1,134 @@
+import org.matrix.android.sdk.api.session.events.model.toValidDecryptedEvent
+import org.matrix.android.sdk.api.session.room.model.message.MessageTextContent
+
+class ValidDecryptedEventTest {
```
Thanks for adding these tests!

mnaturel commented on matrix-sdk-android/src/test/java/org/matrix/android/sdk/internal/database/mapper/EditAggregationSummaryMapperTest.kt:
```diff
@@ -0,0 +1,78 @@
+import org.matrix.android.sdk.internal.database.model.EditionOfEvent
+import org.matrix.android.sdk.internal.database.model.EventEntity
+
+class EditAggregationSummaryMapperTest {
```
Could we rename it `EditAggregatedSummaryEntityMapperTest`? It took me while to find the corresponding tests and IDE does not recognize the associations between the class and the test if naming is not aligned.

BillCarsonFr replied:
ok

mnaturel commented on matrix-sdk-android/src/test/java/org/matrix/android/sdk/internal/database/mapper/EditAggregationSummaryMapperTest.kt:
```diff
@@ -0,0 +1,78 @@
+class EditAggregationSummaryMapperTest {
+
+    @Test
+    fun test() {
```
Should we add a test when we have same timestamp for 2 `EditionOfEvent` to check we take the correct latest event in this situation?

BillCarsonFr replied:
good idea, added

mnaturel reviewed (commented):
Thanks for the updates. I didn't expect android tests (was thinking about manual tests) on database migration but that's awsome we know have tests on this. 

I have just added some small new comments. I wonder also you have seen some of my previous comments (and particularly the ones about `EventEditValidatorTest`)?

bmarty commented:
> I didn't expect android tests (was thinking about manual tests) on database migration but that's awesome we now have tests on this.

Yes, that's nice, @BillCarsonFr can you explain somewhere how to create files like `session_42.realm`

BillCarsonFr commented:
> > I didn't expect android tests (was thinking about manual tests) on database migration but that's awesome we now have tests on this.
> 
> Yes, that's nice, @BillCarsonFr can you explain somewhere how to create files like `session_42.realm`

I launched the app in the emulator (with `LOG_PRIVATE_DATA` to true).
I added some messages to add the entities needed to test my migration.
Then I used the Device File Explorer to export the database `data/data/im.vector.app.debug/files/<hash>/disk_store.realm` and save it on my disk
Then I took the key from logs searching for `Database key for alias  XXX`

And I added that to test assets.

We should create a task to do that for crypto and auth (and probably a better one for session)
The idea would be that it has objects for each table in the base to better test the .transform() on dynamic realm

BillCarsonFr commented:
> Thanks for the updates. I didn't expect android tests (was thinking about manual tests) on database migration but that's awesome we know have tests on this.
> 
> I have just added some small new comments. I wonder also you have seen some of my previous comments (and particularly the ones about `EventEditValidatorTest`)?

I want to handle them in another PR, to rationalize test on IMXCryptoStore and have a bunch of fake events in test.fixtures

mnaturel reviewed (approved):
LGTM.
